inst/include/dust/cuda_launch_control.hpp

```diff
@@ -0,0 +1,233 @@
+    // > which case no problem, just run enough blocks to handle all
+    // > the threads (as normal). If there is a remainder, some of
+    // > last block to will have some idle threads (though these are
+    // > used for the shared mem copy, but don't run a particle).
+    // >
+    // > Here, I found good performance with a block size up to 128,
+    // > dropping off above that. But if you have a relatively large
+    // > number of multi_pars and n_particles_each < 128 it's wasteful
+    // > to leave some many threads idle, so use the nearest multiple
+    // > of 32 up to a max of 128
+    ret.block_size = std::min(static_cast<size_t>(block_size), warp_block_size);
+    ret.block_count =
+      n_pars_effective * (n_particles_each + ret.block_size - 1) /
+      ret.block_size;
+  } else {
+    // If not enough particles per pars to make a whole block use
+    // shared, or if shared_t too big for L1, turn it off, and run
+    // in 'classic' mode where each particle is totally independent
+    set_block_size(ret, block_size, n_particles);
+  }
+
+  return ret;
+}
+
+
+inline launch_control launch_control_simple(size_t block_size, size_t n) {
+  launch_control ret{};
+  set_block_size(ret, block_size, n);
+  return ret;
+}
+
+
+inline launch_control_dust::launch_control_dust() {
+  run = launch_control{};
+  compare = launch_control{};
+  reorder = launch_control{};
+  scatter = launch_control{};
+  index_scatter = launch_control{};
+  interval = launch_control{};
+}
+
+
+inline launch_control_dust::launch_control_dust(const device_config& config,
+                                                size_t n_particles,
+                                                size_t n_particles_each,
+                                                size_t n_state,
+                                                size_t n_state_full,
+                                                size_t n_shared_int,
+                                                size_t n_shared_real,
+                                                size_t real_size,
+                                                size_t data_size) {
+  if (config.enabled_) {
+    const size_t shared_size = config.shared_size_;
+    const size_t run_block_size = config.run_block_size_;
+    run = launch_control_model(n_particles, n_particles_each,
+                               n_shared_int, n_shared_real,
+                               real_size, 0, shared_size, run_block_size);
+    compare = launch_control_model(n_particles, n_particles_each,
+                                   n_shared_int, n_shared_real,
+                                   real_size, data_size, shared_size, 128);
+
+    reorder       = launch_control_simple(128, n_particles * n_state_full);
+    scatter       = launch_control_simple( 64, n_particles * n_state_full);
+    index_scatter = launch_control_simple( 64, n_particles * n_state);
+    interval      = launch_control_simple(128, n_particles);
```

Comment:
Maybe we should make the default block sizes const at the top of the file so that they are easier/clearer to change in future profiling?

Reply:
Although that being said it's quite clear writing them out here too